Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDocumentLayout.mm

```diff
@@ -77,28 +78,76 @@
+            if (isLastPageIndex(pageIndex))
+                return { layoutBoundsForPageAtIndex(pageIndex), { } };
+            return { layoutBoundsForPageAtIndex(pageIndex), layoutBoundsForPageAtIndex(pageIndex + 1) };
+        };
+
+        auto distanceToPage = [documentSpacePoint](const FloatRect& pageLayoutBounds) {
+            if (pageLayoutBounds.contains(documentSpacePoint))
+                return 0.0f;
+
+            auto pointsToCompare = Vector<FloatPoint> { pageLayoutBounds.minXMinYCorner(), pageLayoutBounds.minXMaxYCorner(), pageLayoutBounds.maxXMinYCorner(), pageLayoutBounds.maxXMaxYCorner() };
+
+            bool isAbovePage = documentSpacePoint.y() < pageLayoutBounds.y();
+            bool isBelowPage = documentSpacePoint.y() > pageLayoutBounds.maxY();
+            bool isLeftOfPage = documentSpacePoint.x() < pageLayoutBounds.x();
+            bool isRightOfPage = documentSpacePoint.x() > pageLayoutBounds.maxX();
+
+            bool isWithinPageWidth = isInRange(documentSpacePoint.x(), pageLayoutBounds.x(), pageLayoutBounds.maxX());
+            bool isWithinPageHeight = isInRange(documentSpacePoint.y(), pageLayoutBounds.y(), pageLayoutBounds.maxY());
+
+            if (isAbovePage && isWithinPageWidth)
+                pointsToCompare.append({ documentSpacePoint.x(), pageLayoutBounds.y() });
+            else if (isRightOfPage && isWithinPageHeight)
+                pointsToCompare.append({ pageLayoutBounds.x(), documentSpacePoint.y() });
+            else if (isBelowPage && isWithinPageWidth)
+                pointsToCompare.append({ documentSpacePoint.x(), pageLayoutBounds.maxY() });
+            else if (isLeftOfPage && isWithinPageHeight)
+                pointsToCompare.append({ pageLayoutBounds.x(), documentSpacePoint.y() });
+
+            auto distancesToPoints = WTF::map(pointsToCompare, [documentSpacePoint](FloatPoint point) {
+                return euclidianDistance(point, documentSpacePoint);
+            });
 
-            auto currentPageBounds = layoutBoundsForPageAtIndex(index);
-            if (documentSpacePoint.y() < currentPageBounds.maxY()) {
+            return *std::min_element(distancesToPoints.begin(), distancesToPoints.end());
+        };
 
-                auto pairedPageIndex =  [index, this]() {
-                    if (index % pagesPerRow())
-                        return index - 1;
+        for (PageIndex index = 0; index < pageCount; index += pagesPerRow()) {
+            auto [leftPageLayoutBounds, rightPageLayoutBounds]= layoutBoundsForPairedPages(index);
+
+            if (documentSpacePoint.y() < leftPageLayoutBounds.maxY() || (rightPageLayoutBounds && documentSpacePoint.y() < rightPageLayoutBounds->maxY())) {
+
+                if (leftPageLayoutBounds.contains(documentSpacePoint) || !rightPageLayoutBounds)
+                    return index;
+
+                if (rightPageLayoutBounds->contains(documentSpacePoint))
```

Comment:
Can we not end up accessing the nullopt here? Maybe we should do `rightPageLayoutBounds.value_or({ }).contains(...)` instead?

Reply:
So my idea is that at this point we should have both pages to work with. If we don't then it should have handled it above and return the index of the left page (since the closest page would be the left page if you only have that). I reordered the conditional about to hopefully make that a little more clear!